MSC3550: Allow HTTP 403 as a response to profile lookups (#3550)
matrix-org/matrix-doc: proposals/3550-allow-403-response-profile-lookup.md

> # MSC 3550: Add HTTP 403 to possible profile lookup responses

Reviewer: extreme nitpick: 
```suggestion
# MSC3550: Add HTTP 403 to possible profile lookup responses
```

# Background

> In the current spec, the only response codes listed for  GET /_matrix/client/v3/profile/{userId}

Reviewer: link to the endpoint please :)

(using `v1.1`, not `latest`, as `latest` doesn't help future readers)

are `200` and `404`. However, some servers may not allow profile lookup over federation, and thus
respond to GET /_matrix/client/v3/profile/{userId} with an HTTP 403.

For example, Synapse can be configured to behave in this way by setting:

```
allow_profile_lookup_over_federation=false
```

Thus, this behavior already exists in Synapse, and may cause issues for
clients such as [vector-im/element-web#17269](https://github.com/vector-im/element-web/issues/17269).

# Proposal
The proposal is to allow HTTP 403 as an option for responding to  GET /_matrix/client/v3/profile/{userId}
requests. Allowing HTTP 403 gives clients more specific information as to why a request has 

> failed, thus enabling more precise error handling.

Reviewer: is there an accompanied error code for the 403?

Reviewer 2: Can you include this in the MSC itself?  (I'm assuming`Codes.FORBIDDEN` corresponds to the `M_FORBIDDEN` error code)

Author: Sure no problem!

#MSC3550: Add HTTP 403 to possible profile lookup responses

# Background
In the current spec, the only response codes listed for [GET /_matrix/client/v3/profile/{userId}](https://spec.matrix.org/v1.1/client-server-api/#get_matrixclientv3profileuserid)
are `200` and `404`. However, some servers may not allow profile lookup over federation, and thus
respond to [GET /_matrix/client/v3/profile/{userId}](https://spec.matrix.org/v1.1/client-server-api/#get_matrixclientv3profileuserid) with an HTTP 403.

For example, Synapse can be configured to behave in this way by setting:

```
allow_profile_lookup_over_federation=false
```

Thus, this behavior already exists in Synapse, and may cause issues for
clients such as [vector-im/element-web#17269](https://github.com/vector-im/element-web/issues/17269).

# Proposal
The proposal is to allow HTTP 403 as an option for responding to [GET /_matrix/client/v3/profile/{userId}](https://spec.matrix.org/v1.1/client-server-api/#get_matrixclientv3profileuserid)

> requests. Allowing HTTP 403 gives clients more specific information as to why a request has 
> failed, thus enabling more precise error handling. For example, Synapse currently 
> returns the `M_FORBIDDEN` error code along with the HTTP 403 response.  

Reviewer: ```suggestion
requests. Allowing HTTP 403 gives clients more specific information as to why a request has 
failed, thus enabling more precise error handling. The 403 would be accompanied by an
`M_FORBIDDEN` error code. 
```
